Update state version and workspace API docs (#379)
hashicorp/terraform-docs-common: website/docs/cloud-docs/api-docs/state-versions.mdx

---------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
`hosted-json-state-download-url` | A URL from which you can download the state data in a [stable format](/terraform/internals/json-format) appropriate for external integrations to consume. Only available if the state was created by Terraform 1.3+.
`hosted-state-download-url`      | A URL from which you can download the raw state data, in the format used internally by Terraform.

> `hosted-json-state-upload-url`   | A URL to which you can upload the state data, in a [stable format](/terraform/internals/json-format) appropriate for external integrations to consume. You can only upload JSON state content once per state version.

Reviewer: ```suggestion
`hosted-json-state-upload-url`   | A URL to which you can upload state data in a [stable format](/terraform/internals/json-format) appropriate for external integrations to consume. You can upload JSON state content once per state version.
```

> `hosted-json-state-upload-url`   | A URL to which you can upload the state data, in a [stable format](/terraform/internals/json-format) appropriate for external integrations to consume. Only available if the JSON state content has not been uploaded.

Reviewer 2: regarding the last sentence, what would do the uploading? is this only an operator, or would TFC do this for you? and how do they generate the state data to upload?

Author: Yup this is a little tricky to word! We can clean this up a bit. Essentially there are two ways to upload: either by including the content in request body for `Create State Version` or, by omitting the content in that request and uploading the content to the URL provided in the response body `hosted-state-upload-url` and `hosted-json-state-upload-url`.

> `hosted-json-state-upload-url`   | A URL from which you can upload the state data to, in a [stable format](/terraform/internals/json-format) appropriate for external integrations to consume. Only available if the JSON state content has not been uploaded.
> `hosted-state-upload-url`        | A URL from which you can upload the state data to, in the format used internally by Terraform. Only available if the state content has not been uploaded.

Reviewer 3: ```suggestion
`hosted-json-state-upload-url`   | A URL to which you can upload the state data, in a [stable format](/terraform/internals/json-format) appropriate for external integrations to consume. Only available if the JSON state content has not been uploaded.
`hosted-state-upload-url`        | A URL to which you can upload the state data, in the format used internally by Terraform. Only available if the state content has not been uploaded.
```

`modules`                        | Extracted information about the Terraform modules in this state data. Populated asynchronously.
`providers`                      | Extracted information about the Terraform providers used for resources in this state data. Populated asynchronously.

> `provisional`                    | A boolean flag that denotes a state version that was created but has not yet marked as the current version for a workspace.

Reviewer 2: should this be state "version", or state "snapshot"? i worry state version is a little loaded since we have that field in the state file also

Author: We are avoiding using the term "snapshot", because a snapshot is indistinguishable from a state version. The key difference is that a "snapshot" is a state version that has not been set as the current version for the workspace. Rather than introduce a different term altogether, we decided to use the term "provisional" state version.

> `provisional`                    | A boolean flag that denotes a state version that was created but has not yet marked as the current version for a workspace.

Reviewer 2: ```suggestion
`provisional`                    | A boolean flag indicating that Terraform Cloud created a state version but did not yet finalize it and mark it as the current version for the workspace.
```

Author: The only thing I'll omit here is the "not yet finalize it". Provisional state versions can be finalized.

> `provisional`                    | A boolean flag that denotes a state version that was created but has not yet marked as the current version for a workspace.

Reviewer 4: This is nice

`resources`                      | Extracted information about the resources in this state data. Populated asynchronously.
`resources-processed`            | A Boolean flag indicating whether Terraform Cloud has finished asynchronously extracting outputs, resources, and other information about this state data.
`serial`                         | The serial number of this state instance, which increases every time Terraform creates new state in the workspace.
`state-version`                  | The version of the internal state format used for this state. Different Terraform versions read and write different format versions, but it only changes infrequently.

> `status`                         | Indicates a state version's content upload status, can be `pending`, `finalized` or `discarded`.

Reviewer 2: same as above - should "version" be "snapshot"? (or something else?)

> `status`                         | Indicates a state version's content upload status, can be `pending`, `finalized` or `discarded`.

Reviewer 2: ```suggestion
`status`                         | Indicates a state version's content upload status. Can be `pending`, `finalized` or `discarded`.
```

`terraform-version`              | The Terraform version that created this state. Populated asynchronously.
`vcs-commit-sha`                 | The SHA of the configuration commit used in the Terraform run that produced this state. Only present if the workspace is connected to a VCS repository.

> `vcs-commit-url`                 | A link to the configuration commit used in the Terraform run that produced this state. Only present if the workspace is connected to a VCS repository.

Reviewer 4: Duplicated from my PR #421 

```suggestion
`vcs-commit-url`                 | A link to the configuration commit used in the Terraform run that produced this state. Only present if the workspace is connected to a VCS repository.

### State Version States

The state version state is found in `data.attributes.status`, and you can reference the following list of possible states.

A state version created through the API or CLI will only be listed in the UI if it is in the `finalized` state.

| State        | Description                                                                                                                                                                                                                                                                                                                                                                                                                        |
| ------------ | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `pending`    | The initial status of a state version after it has been created without encoding the state data within the request. Pending state versions do not contain state data and will not appear in the UI. The workspace cannot be unlocked normally until the latest state version has been finalized.                                     |
| `finalized`  | The state version has been uploaded successfully to Terraform Cloud or if the state version was created with a valid `state` attribute.                                                                                                                                                                                                              |
| `discarded` | The state version has been discarded because it was provisional and not promoted to be the current state version of the workspace. |
```

Reviewer: Possibly silly question: Can we call them something besides "state version states"? Reading the first sentence is kinda confusing with the number of times the word "state" comes up.

Reviewer 4: For sure. You are right about that. I got a little carried away! State Version Status would be the next best thing

| `data.type`                          | string  |           | Must be `"state-versions"`.                                                                                                                                                                                                                                                                |
| `data.attributes.serial`             | integer |           | The serial of the state version. Must match the serial value extracted from the raw state file.                                                                                                                                                                                            |
| `data.attributes.md5`                | string  |           | An MD5 hash of the raw state version.                                                                                                                                                                                                                                                       |

> | `data.attributes.state`              | string  | (nothing) | **Optional** Base64 encoded raw state file. If not included in this request, must be uploaded to `hosted-state-upload-url`.                                                                                                                                                                                                                                                              |

Reviewer 4: ```suggestion
| `data.attributes.state`              | string  | (nothing) | **Optional** Base64 encoded raw state file. If not included in this request, `status` will be `pending` until state is uploaded to `hosted-state-upload-url`.                                                                                                                                                                                                                                                             |
```

> | `data.attributes.state`              | string  | (nothing) | **Optional** Base64 encoded raw state file. If not included in this request, `status` will be `pending` until state is uploaded to `hosted-state-upload-url`.                                                                                                                                                                                                                                                              |

Reviewer 4: ```suggestion
| `data.attributes.state`              | string  | (nothing) | **Optional** Base64 encoded raw state file. If omitted, you must use the upload method below to complete the state version creation. The workspace may not be unlocked normally until the state version is uploaded.                                                                                                                                                                                                                                                              |
```

}
```

## Upload State and JSON State

> -> **Note:** While you can upload state version content in the same request when creating a state version, we strongly recommend that you upload content separately.

Reviewer: We try to minimize the number of callouts in the docs, if possible. Does this need to be a call out? It seems like this would get enough attention from being the first sentence:
```suggestion
You can upload state version content in the same request when creating a state version. However, we _strongly_ recommend that you upload content separately.
```
Also, why do we recommend this? Can we give the users more context?

> -> **Note:** While you can upload state version content in the same request when creating a state version, it is strongly recommended you upload content directly to the provided URL.

Reviewer 3: ```suggestion
-> **Note:** While you can upload state version content in the same request when creating a state version, we strongly recommend that you upload content to the provided URL after creating the state version.
```

(Flow of events is a little less clear without that.)

> `PUT https://archivist.terraform.io/v1/object/<UNIQUE OBJECT ID>`

Reviewer 2: do they need to get the object id? where do they reference the URL?

Author: The entire URL is provided `hosted-state-upload-url` or `hosted-json-state-upload-url` field, so they don't need to fetch any object ID.

> **The URL is the `hosted-state-upload-url` or `hosted-json-state-upload-url` returned when you create a `state-version`. Once you upload state content, the URL is hidden on the resource and no longer available.**

Reviewer: We don't bold for emphasis in the docs, we use italics. If this is important for folks to know and needs more emphasis, we could add a call out?
```suggestion
Terraform Cloud returns a `hosted-state-upload-url` or `hosted-json-state-upload-url` when you create a `state-version`. Once you upload state content, this URL is hidden on the resource and _no longer available_.
```

> -> **Note:** While you can upload state version content in the same request when creating a state version, we strongly recommended that you upload content directly to the provided URL.
> 
> `PUT https://archivist.terraform.io/v1/object/<UNIQUE OBJECT ID>`
> 
> **The URLs are provided in the `hosted-state-upload-url` and `hosted-json-state-upload-url` attributes when creating a `state-version` resource. Once uploaded, the URL is hidden on the resource and no longer available.**

Reviewer 2: ```suggestion
-> **Note:** While you can upload state version content in the same request when creating a state version, we strongly recommend that you upload content separately.

`PUT https://archivist.terraform.io/v1/object/<UNIQUE OBJECT ID>`

**The URL is the `hosted-state-upload-url` or `hosted-json-state-upload-url` returned when you create a `state-version`. Once you upload state content, the URL is hidden on the resource and no longer available.**
```

### Sample Request

> **@filename is the name of Terraform statefile you wish to upload.**

Reviewer 4: ```suggestion
**@filename is the name of Terraform state file you wish to upload.**
```

This endpoint supports pagination [with standard URL query parameters](/terraform/cloud-docs/api-docs#query-parameters). Remember to percent-encode `[` as `%5B` and `]` as `%5D` if your tooling doesn't automatically encode URLs.

| Parameter                    | Description                                                                        |
|------------------------------|------------------------------------------------------------------------------------|
| `filter[workspace][name]`    | **Required** The name of one workspace to list versions for.                       |
| `filter[organization][name]` | **Required** The name of the organization that owns the desired workspace.         |

> | `filter[status]`             | **Optional.** If omitted, the endpoint will return all `finalized` state versions. |

Reviewer 4: I don't think this has a default

```suggestion
| `filter[status]`             | **Optional.** Filter state versions by status: `pending`, `finalized`, or `discarded` |
```

                "created-at": "2021-06-08T01:22:03.794Z",
                "size": 940,
                "hosted-state-download-url": "https://archivist.terraform.io/v1/object/...",

>                 "hosted-state-upload-url": "https://archivist.terraform.io/v1/object/...",

Reviewer 4: If a state is finalized (or if it has a download URL), it won't have this upload URL in the response 🤔 

```suggestion
                "hosted-state-upload-url": null,
```

                "hosted-json-state-download-url": "https://archivist.terraform.io/v1/object/...",

>                 "hosted-json-state-upload-url": "https://archivist.terraform.io/v1/object/...",

Reviewer 4: ```suggestion
                "hosted-json-state-upload-url": null,
```

                "created-at": "2021-06-01T21:40:25.941Z",
                "size": 819,
                "hosted-state-download-url": "https://archivist.terraform.io/v1/object/...",

>                 "hosted-state-upload-url": "https://archivist.terraform.io/v1/object/...",

Reviewer 4: ```suggestion
                "hosted-state-upload-url": null,
```

                "hosted-json-state-download-url": "https://archivist.terraform.io/v1/object/...",

>                 "hosted-json-state-upload-url": "https://archivist.terraform.io/v1/object/...",

Reviewer 4: ```suggestion
                "hosted-json-state-upload-url": null,
```
